subprojects/model-core/src/main/java/org/gradle/internal/instantiation/generator/Jsr330ConstructorSelector.java

```diff
@@ -50,17 +50,18 @@ public void vetoParameters(ClassGenerator.GeneratedConstructor<?> constructor, O
 
     @Override
     public <T> ClassGenerator.GeneratedConstructor<? extends T> forType(final Class<T> type) throws UnsupportedOperationException {
-        CachedConstructor constructor = constructorCache.get(type, () -> {
+        // We cache only constructor index, so we don't keep reference to implClass in constructor cache and GC can claim it faster
+        ClassGenerator.GeneratedClass<?> implClass = classGenerator.generate(type);
```

Comment:
Instead of using indexes here, wouldn't it be better to couple the lifetime of the constructor cache with the lifetime of the `GeneratedClass`? Since as soon as there isn't any reference to the generated class anymore, then the selected constructor isn't used anymore.

Reply:
I can try. So having implClass as a key of constructorCache and still returning constructor instead of index right?

Reply:
Yeah, something like that. I was wondering if we could tie this information directly to the things in `AbstractClassGenerator.generatedClasses`. Though there shouldn't be a big difference now, so I don't understand why the elements in the original cache (before your changes) can't be collected in the same way as they are for the generated classes themselves.

The classes should actually be collected the same way according to the documentation on `CrossBuildInMemoryCacheFactory.newClassMap()`, since only the keys matter, not the values. Do you have an idea what is going on there?

Reply:
So problematic classes are one created with worker API and that is why worker api isolated classloader is not gced.

This is one of strong references chain that keeps `FooAction$Inject` and worker api classloader from getting gced. 
<img width="1529" alt="Screenshot 2021-09-22 at 19 58 25" src="https://user-images.githubusercontent.com/3939893/134396937-86358da3-2b2e-44a9-a718-05c5efae536d.png">

So I see that there is also reference from `valuesForPreviousSession` inside `DefaultCrossBuildInMemoryCache`. So actually caching previous session values and keeping strong reference to them is expected (also I see documentation on `DefaultCrossBuildInMemoryCacheFactory` states that).

Unfortunately that also affects worker api classloader not being cleaned up, but also I think then my solution is not correct in that case.